src/protocols/secure_channel/CASESession.cpp

```diff
@@ -479,67 +524,82 @@ CHIP_ERROR CASESession::SendSigmaR2()
-        bbuf.Put(mRemotePubKey, mRemotePubKey.Length());
-
-        VerifyOrExit(bbuf.Fit(), err = CHIP_ERROR_NO_MEMORY);
+        System::PacketBufferTLVWriter tlvWriter;
+        TLV::TLVType outerContainerType = TLV::kTLVType_NotSpecified;
+
+        tlvWriter.Init(std::move(msg_R2_Signed));
+        SuccessOrExit(err = tlvWriter.StartContainer(TLV::AnonymousTag, TLV::kTLVType_Structure, outerContainerType));
+        SuccessOrExit(err = tlvWriter.PutBytes(CASETLVTag::kResponderEphPubKey, mEphemeralKey.Pubkey(),
+                                               static_cast<uint32_t>(mEphemeralKey.Pubkey().Length())));
+        SuccessOrExit(err = tlvWriter.PutBytes(CASETLVTag::kNOC, mOpCredSet->GetDevOpCred(mTrustedRootId),
+                                               mOpCredSet->GetDevOpCredLen(mTrustedRootId)));
+        SuccessOrExit(err = tlvWriter.PutBytes(CASETLVTag::kInitiatorEphPubKey, mRemotePubKey,
+                                               static_cast<uint32_t>(mRemotePubKey.Length())));
+        SuccessOrExit(err = tlvWriter.EndContainer(outerContainerType));
+        SuccessOrExit(err = tlvWriter.Finalize(&msg_R2_Signed));
     }
 
     // Step 7
-    err = mOpCredSet->SignMsg(mTrustedRootId, msg_R2_Signed.Get(), msg_r2_signed_len, sigmaR2Signature);
+    err = mOpCredSet->SignMsg(mTrustedRootId, msg_R2_Signed->Start(), msg_R2_Signed->DataLength(), sigmaR2Signature);
     SuccessOrExit(err);
 
     // Step 8
     msg_r2_signed_enc_len = sizeof(uint16_t) + mOpCredSet->GetDevOpCredLen(mTrustedRootId) + sigmaR2Signature.Length();
 
-    VerifyOrExit(msg_R2_Encrypted.Alloc(msg_r2_signed_enc_len), err = CHIP_ERROR_NO_MEMORY);
-    {
-        Encoding::LittleEndian::BufferWriter bbuf(msg_R2_Encrypted.Get(), msg_r2_signed_enc_len);
-
-        bbuf.Put16(mOpCredSet->GetDevOpCredLen(mTrustedRootId));
-        bbuf.Put(mOpCredSet->GetDevOpCred(mTrustedRootId), mOpCredSet->GetDevOpCredLen(mTrustedRootId));
-        bbuf.Put(sigmaR2Signature, sigmaR2Signature.Length());
+    msg_R2_Encrypted = System::PacketBufferHandle::New(msg_r2_signed_enc_len + System::PacketBuffer::kDefaultHeaderReserve);
+    VerifyOrExit(!msg_R2_Encrypted.IsNull(), err = CHIP_SYSTEM_ERROR_NO_MEMORY);
 
-        VerifyOrExit(bbuf.Fit(), err = CHIP_ERROR_NO_MEMORY);
+    // Generate Sigma2 TBE Data
+    {
+        System::PacketBufferTLVWriter tlvWriter;
+        TLV::TLVType outerContainerType = TLV::kTLVType_NotSpecified;
+
+        tlvWriter.Init(std::move(msg_R2_Encrypted));
+        SuccessOrExit(tlvWriter.StartContainer(TLV::AnonymousTag, TLV::kTLVType_Structure, outerContainerType));
+        SuccessOrExit(err = tlvWriter.PutBytes(CASETLVTag::kNOC, mOpCredSet->GetDevOpCred(mTrustedRootId),
+                                               mOpCredSet->GetDevOpCredLen(mTrustedRootId)));
+        SuccessOrExit(
+            err = tlvWriter.PutBytes(CASETLVTag::kSignature, sigmaR2Signature, static_cast<uint32_t>(sigmaR2Signature.Length())));
```

Comment:
According to the spec: Section `4.12.2.5. Signature` the signature should be encoded in its raw 64 representation over the wire.

Reply:
We will address this in a separate commit where we'll bring the protocol up to the current spec. The intention for this specific PR is just to add TLV format.